### DISCUSSION TITLE
Some way to get unittest.TestCase.assertIsInstance to type narrow?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5303] [Q&A] [answered]
For the builtin [`unittest.TestCase.assertIsInstance`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIsInstance):

```python
import unittest
from typing import Optional


def get_int() -> Optional[int]:
    return 1


class TestA(unittest.TestCase):
    def test_a(self):
        x = get_int()
        self.assertIsInstance(x, int)
        # assert isinstance(x, int)  # feels redundant
        result = x + x  # Operator "+" not supported for types "int | None" and "int | None"

```

Is there a way to define a `TypeGuard` for this (e.g. using `NoReturn`)? Or is there no such existing style of `TypeGuard` in Python? I'm kind of at a loss how to define a `TypeGuard` where the guard is implicit in the continued execution past the statement rather than a bool that can be used as part of a condition to narrow the type.

**Comment by erictraut** [comment]
I don't think there's a way to do this in Python.

Doesn't `assert isinstance(x, int)` work fine with the `unittest` library? I thought that it was able to use exception handlers and introspection to handle `assert isinstance(x, int)` the same as `self.assertIsInstance(x, int)`. Maybe I'm confusing it with `pytest`.

**Reply by ringohoffman** [reply]
Hmm I feel surprised by this behavior:

```python
from typing import Any, NoReturn, Type, TypeVar, Union, overload

T = TypeVar("T")


@overload
def assertIsInstance(obj: T, class_or_tuple: Type[T]) -> None:
    ...


@overload
def assertIsInstance(obj: T, class_or_tuple: Type[object]) -> NoReturn:
    ...


def assertIsInstance(obj: T, class_or_tuple: Union[Type[T], Type[object]]) -> Union[None, NoReturn]:
    if not isinstance(obj, class_or_tuple):
        raise AssertionError(f"{obj} is not an instance of {class_or_tuple}")
    return None


x = assertIsInstance(1, int)
reveal_type(x)  # Type of "x" is "None"

y = assertIsInstance(1, str)
reveal_type(y)  # Type of "y" is "None"

a = assertIsInstance("abc", int)
reveal_type(a)  # Type of "a" is "None"


def raiseError(obj: Any, class_or_tuple: Any) -> NoReturn:
    raise AssertionError(f"{obj} is not an instance of {class_or_tuple}")

z = raiseError(1, str)
reveal_type(z)  # Code is unreachable

```

Am I setting up my overload signatures wrong? Wouldn't doesn't the first overload signature only get matched when `obj` is an instance of `class_or_tuple`?

**Reply by erictraut** [reply, marked as answer]
The constraint solver is free to "solve" the type variable by creating a union. It doesn't require that the types of the arguments passed to `obj: T` and `class_or_tuple: type[T]` be related in any way.

```python
def foo(a: T, b: T) -> T:
    ...

reveal_type(foo(1, "")) # pyright: int | str, mypy: object
```

As I said, I don't think there's a way to do what you're trying to do here.
